### PR TITLE
feat: Support preloading container images into Bottlerocket data volumes with Karpenter    

### PR DESCRIPTION
**Pull request by lindarr915**
### What does this PR do?

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

### Motivation

Reduce the cold start time for ray worker pods. 
By preloading container images into the bottlerocket OS data volume, ~10 mins can be saved for container startup time as it is not necessary to download and extract the container images layers.    

https://github.com/awslabs/data-on-eks/issues/559

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [ ] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

**Review by askulkarni2 — COMMENTED** [review]
@lindarr915 thanks for the PR. Approach looks good but lets fix the karpenter helm chart to accept the snapshot_id instead of introducing a breaking change this way.

**Comment by lindarr915** [comment]
Move to draft

**Comment by lindarr915** [comment]
Open another PR to merge from another branch

File: ai-ml/jark-stack/preload-container-image-ami/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# Preload Container Images
```
Comment (askulkarni2) [inline ai-ml/jark-stack/preload-container-image-ami/README.md:1]: Let's kill this doc, you are documenting it in the website which is good enough.

File: ai-ml/jark-stack/terraform/README.md [hunk 2]
```diff
@@ -1,6 +1,79 @@
 # JupyterHub, Argo, Ray, Kubernetes
 
-Docs coming soon...
+# Prerequisites
```
Comment (askulkarni2) [inline ai-ml/jark-stack/terraform/README.md:3]: Lets remove this doc as well. We have an open [issue](https://github.com/awslabs/data-on-eks/issues/445) to replace the existing Ray on EKS blueprint with JARK stack along with the  documentation in the website.

File: ai-ml/jark-stack/terraform/addons.tf [hunk 3]
```diff
@@ -145,9 +145,10 @@ module "eks_blueprints_addons" {
 #---------------------------------------------------------------
 # Data on EKS Kubernetes Addons
 #---------------------------------------------------------------
+
 module "data_addons" {
   source  = "aws-ia/eks-data-addons/aws"
-  version = "~> 1.31.4" # ensure to update this to the latest/desired version
+  version = "~> 1.40"   # source  = "aws-ia/eks-data-addons/aws"
```
Comment (askulkarni2) [inline ai-ml/jark-stack/terraform/addons.tf:151]:
```suggestion
  version = "~> 1.40"
```

File: ai-ml/jark-stack/terraform/install.sh [hunk 4]
```diff
@@ -6,6 +6,10 @@ targets=(
   "module.eks"
 )
 
+# Get the snapshot ID with ./snapshot.sh
+# Uncomment the command line below and replace the value with the snapshot ID
+# export TF_VAR_bottlerocket_data_disk_snpashot_id=snap-0ffe58efdd845d938
+
```
Comment (askulkarni2) [inline ai-ml/jark-stack/terraform/install.sh:12]:
```suggestion
```

We can probably remove this as this is covered in the doc.

File: ai-ml/jark-stack/terraform/karpenter.tf [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    effect    = "Allow"
+    sid       = "Karpenter"
+  }
+}
+
+resource "aws_iam_role_policy_attachment" "karpenter_controller_policy_attachment" {
+  # name       = "karpenter-controller-policy-attachment"
+  role      = module.eks_blueprints_addons.karpenter.iam_role_name
+  policy_arn = aws_iam_policy.karpenter_controlloer_policy.arn
+}
```
Comment (askulkarni2) [inline ai-ml/jark-stack/terraform/karpenter.tf:22]: Move this to `addons.tf` and provide the policy_document as value for the `karpenter.source_policy_document`. The add-on will merge this policy into the default policy. See https://github.com/aws-ia/terraform-aws-eks-blueprints-addons/blob/c003cad83f930a16395a5e4cddd7daa1d74c76a3/main.tf#L2802 for reference.

File: website/docs/gen-ai/inference/stablediffusion-gpus.md [hunk 6]
```diff
@@ -6,10 +6,11 @@ import CollapsibleContent from '../../../src/components/CollapsibleContent';
 
 :::info
 
-We are actively enhancing this blueprint to incorporate improvements in observability and logging.
-
+- We are actively enhancing this blueprint to incorporate improvements in observability and logging.
+- To accelerate the deployment of image retrieval on Ray workerss, refer to [Preload container images into Bottlerocket data volumes with Karpenter with EBS Snapshots](../../bestpractices/preload-container-images)
```
Comment (askulkarni2) [inline website/docs/gen-ai/inference/stablediffusion-gpus.md:10]:
```suggestion
- To accelerate the deployment of image retrieval on Ray workers, refer to [Preload container images into Bottlerocket data volumes with Karpenter with EBS Snapshots](../../bestpractices/preload-container-images)
```

File: website/docs/bestpractices/preload-container-images.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  blockDeviceMappings:
+    - deviceName: /dev/xvdb
+      ebs:
+        volumeSize: 150Gi
+        volumeType: gp3
+        kmsKeyID: "arn:aws:kms:us-west-2:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab" # Specify KMS ID if you use custom KMS key
+        snapshotID: snap-0123456789 # Specify your snapshot ID here
+```
+
+# End-to-End deployment example
```
Comment (askulkarni2) [inline website/docs/bestpractices/preload-container-images.md:91]: Merge this section into the Stable-diffusion example as an optional section.

File: ai-ml/jark-stack/preload-container-image-ami/README.md (hunk 1 above)
Comment (lindarr915) [reply ai-ml/jark-stack/preload-container-image-ami/README.md:1]: No problem

**Comment by lindarr915** [comment]
@askulkarni2 Feel free to to review again

**Review by askulkarni2 — APPROVED** [review]
LGTM! 🥳

**Pull request merged**